### PR TITLE
Add Value Object and Exceptions to migrate Contact>Store in BO

### PR DESCRIPTION
**Pull request by LouisAUTHIE**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Adds Value Object and a few exceptions in order to migrate Contact>Store in BO
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | N/A
| How to test?      | N/A
| Possible impacts? | N/A

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27358)

### CODE REVIEW

File: src/Core/Domain/Store/Exception/StoreConstraintException.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Is thrown when store is invalid
+ */
+class StoreConstraintException extends StoreException
+{
+	/**
+	 * Thrown when provided store id is not valid
+	 */
+	public const INVALID_ID = 10;
+}
```
Comment (Progi1984) [inline src/Core/Domain/Store/Exception/StoreConstraintException.php]:
```suggestion
}

```

File: src/Core/Domain/Store/Exception/StoreException.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Store\Exception;
+
+use PrestaShop\PrestaShop\Core\Domain\Exception\DomainException;
+
+class StoreException extends DomainException
+{
+}
```
Comment (Progi1984) [inline src/Core/Domain/Store/Exception/StoreException.php]:
```suggestion
}

```

File: src/Core/Domain/Store/Exception/StoreNotFoundException.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Store\Exception;
+
+/**
+ * Thrown when store was not found
+ */
+class StoreNotFoundException extends StoreException
+{
+}
```
Comment (Progi1984) [inline src/Core/Domain/Store/Exception/StoreNotFoundException.php]:
```suggestion
}

```

File: src/Core/Domain/Store/ValueObject/StoreId.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	/**
+	 * @return int
+	 */
+	public function getValue(): int
+	{
+		return $this->storeId;
+	}
+}
```
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/StoreId.php]:
```suggestion
}

```

File: src/Core/Domain/Store/ValueObject/StoreId.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	private $storeId;
+
+	/**
+	 * @param int $storeId
+	 */
+	public function __construct(int $storeId)
+	{
+		if (0 >= $storeId) {
+			throw new StoreConstraintException(sprintf('Invalid store id "%d"', $storeId), StoreConstraintException::INVALID_ID);
+		}
```
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/StoreId.php]: Could you add an unit test for testing the exception ?

**Review by Progi1984 — COMMENTED** [review]
@LouisAUTHIE Thank you for your contribution. Small feedbacks

**Comment by kpodemski** [comment]
If it's a part of the migration, why not add it to the related Pull Request? 🤔

**Comment by Progi1984** [comment]
@kpodemski It's a willing. Dividing in small pull requests more readable, more reviewable and easy mergeable.

File: src/Core/Domain/Store/ValueObject/StoreId.php (hunk 5 above)
Comment (matthieu-rolland) [reply src/Core/Domain/Store/ValueObject/StoreId.php]: You will find examples of how to test an exception in existing unit tests, for example here: `tests/Unit/Adapter/Image/ImageValidatorTest.php` line 63

it's pretty straightforward 👍

**Review by matthieu-rolland — COMMENTED** [review]
lgtm (waiting for unit test before aprove)

**Review by matthieu-rolland — DISMISSED** [review]
LGTM 👍

**Comment by matthieu-rolland** [comment]
CS Fixer is not happy it seems...

you can fix this with this command, from the root of your project:

`./vendor/friendsofphp/php-cs-fixer/php-cs-fixer fix src/Core/Domain/Store/ValueObject/StoreId.php`

then commit and push the modification it will make on the file 👍

**Comment by LouisAUTHIE** [comment]
Indentation problem fixed thanks for the reviews

**Review by Progi1984 — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by WebHelpersPau** [comment]
Should be ok now. Thanks for the reviews

File: tests/Unit/Core/Domain/Store/ValueObject/StoreIdTest.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $this->expectException(StoreConstraintException::class);
+        $storeId = new StoreId(-3);
+    }
+
+    public function testItThrowsExceptionWhenStoreIdEqualsZero(): void
+    {
+        $this->expectException(StoreConstraintException::class);
+        $storeId = new StoreId(0);
+    }
```
Comment (sowbiba) [inline tests/Unit/Core/Domain/Store/ValueObject/StoreIdTest.php:47]: missing success test

**Review by sowbiba — COMMENTED** [review]
Just a small feedback
And can you please do the commit messages in English next time ? ;)

**Review by sowbiba — COMMENTED** [review]
Just a small feedback
And can you please do the commit messages in English next time ? ;)

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
thank you @LouisAUTHIE !

**Pull request merged**